recommenders/models/ncf/dataset.py

```diff
@@ -1,321 +1,450 @@
-            )
-
-            # get negative pools for every user based on training and test interactions
-            test_interact_status = pd.merge(
-                test_interact_status, self.interact_status, on=self.col_user, how="left"
-            )
-            test_interact_status[
-                self.col_item + "_negative"
-            ] = test_interact_status.apply(
-                lambda row: row[self.col_item + "_negative"]
-                - row[self.col_item + "_interacted_test"],
-                axis=1,
-            )
-            test_ratings = pd.merge(
-                self.test,
-                test_interact_status[[self.col_user, self.col_item + "_negative"]],
-                on=self.col_user,
-                how="left",
-            )
-
-            # sample n_neg_test negative samples for testing
-            try:
-                test_ratings[self.col_item + "_negative"] = test_ratings[
-                    self.col_item + "_negative"
-                ].apply(lambda x: random.sample(x, self.n_neg_test))
-
-            except Exception:
-                min_num = min(map(len, list(test_ratings[self.col_item + "_negative"])))
-                warnings.warn(
-                    "n_neg_test is larger than negative items set size! We will set n_neg as the smallest size: %d"
-                    % min_num
-                )
-                test_ratings[self.col_item + "_negative"] = test_ratings[
-                    self.col_item + "_negative"
-                ].apply(lambda x: random.sample(x, min_num))
-
-            self.test_data = []
-
-            # generate test data
-            for row in test_ratings.itertuples():
-                self.test_users, self.test_items, self.test_ratings = [], [], []
-
-                self.test_users.append(int(getattr(row, self.col_user)))
-                self.test_items.append(int(getattr(row, self.col_item)))
-                self.test_ratings.append(float(getattr(row, self.col_rating)))
-
-                for i in getattr(row, self.col_item + "_negative"):
-                    self.test_users.append(int(getattr(row, self.col_user)))
-                    self.test_items.append(int(i))
-                    self.test_ratings.append(float(0))
-
-                self.test_data.append(
-                    [
-                        [self.id2user[x] for x in self.test_users],
-                        [self.id2item[x] for x in self.test_items],
-                        self.test_ratings,
-                    ]
+                user_negative_examples = self._get_negative_examples(
+                    user, user_negative_item_pool, n_samples
                 )
+                user_examples = pd.concat([user_positive_examples, user_negative_examples])
+                shuffle_buffer.append(user_examples)
+                shuffle_buffer_len = sum([df.shape[0] for df in shuffle_buffer])
+                if (shuffle_buffer_len >= shuffle_size):
```

Comment:
In this code `shuffle_buffer` reads data in the same order as users, right? So, if `shuffle_size` is small enough the training epoch will see the first users only according to the ordering of `train_datafile.users`. As a result the last users will never be used, will they? So I think you need to permute `train_datafile.users` in line 413.

Reply:
Yes the `train_loader` always reads data in the same order, adding users' data to shuffle buffer. If shuffle buffer is small, it will get shuffled and released more often, but the buffer will always contain all the data for at least one user. After being released, the shuffle buffer will be initialized with any remainder following the last release and start appending next set of users' data. Line 433 will release any remaining data if there is still data in the shuffle buffer after iterating through all users. The only problem with small `shuffle_size` is that the quality of shuffle will be poor because the same users are likely to appear in the same batch over every epoch.

Reply:
I see. It looks correct then.